packages/ozone/src/scheduled-action/service.ts

```diff
@@ -0,0 +1,276 @@
+          .orWhere('executeUntil', '<=', endTime.toISOString())
+          .orWhere((sqb) => {
+            return sqb
+              .where('executeUntil', 'is', null)
+              .where('executeAfter', '<=', endTime.toISOString())
+          })
+      })
+    }
+
+    if (cursor) {
+      query = query.where(
+        'id',
+        direction === 'asc' ? '>' : '<',
+        parseInt(cursor, 10),
+      )
+    }
+
+    const actions = await query.orderBy('id', direction).limit(limit).execute()
+
+    return {
+      actions,
+      cursor: actions.at(-1)?.id?.toString(),
+    }
+  }
+
+  async cancelScheduledActions(subjects: string[]): Promise<{
+    succeeded: string[]
+    failed: { did: string; error: string; errorCode?: string }[]
+  }> {
+    const succeeded: string[] = []
+    const failed: { did: string; error: string; errorCode?: string }[] = []
+
+    for (const did of subjects) {
+      try {
+        const result = await this.db.db
+          .updateTable('scheduled_action')
+          .set({
+            status: 'cancelled',
+            updatedAt: new Date().toISOString(),
+          })
+          .where('did', '=', did)
+          .where('status', '=', 'pending')
+          .executeTakeFirst()
+
+        if (result.numUpdatedRows && result.numUpdatedRows > 0) {
+          succeeded.push(did)
+        } else {
+          failed.push({
+            did,
+            error: 'No pending scheduled actions found for subject',
+            errorCode: 'NoPendingActions',
+          })
+        }
+      } catch (error) {
+        failed.push({
+          did,
+          error: error instanceof Error ? error.message : 'Unknown error',
+          errorCode: 'DatabaseError',
+        })
+      }
+    }
+
+    return { succeeded, failed }
+  }
+
+  async getPendingActionsToExecute(
+    now: Date,
+  ): Promise<Selectable<ScheduledAction>[]> {
+    return await this.db.db
+      .selectFrom('scheduled_action')
+      .selectAll()
+      .where('status', '=', 'pending')
+      .where((qb) => {
+        return qb
+          .orWhere('executeAfter', '<=', now.toISOString())
+          .orWhere('executeAt', '<=', now.toISOString())
+      })
+      .execute()
+  }
+
+  async markActionAsExecuted(
+    actionId: number,
+    executionEventId: number,
+  ): Promise<void> {
+    await this.db.db
+      .updateTable('scheduled_action')
+      .set({
+        status: 'executed',
+        lastExecutedAt: new Date().toISOString(),
+        executionEventId,
+        updatedAt: new Date().toISOString(),
+      })
+      .where('id', '=', actionId)
+      .execute()
+  }
+
+  async markActionAsFailed(
+    actionId: number,
+    failureReason: string,
+  ): Promise<void> {
+    await this.db.db
+      .updateTable('scheduled_action')
+      .set({
+        status: 'failed',
+        lastExecutedAt: new Date().toISOString(),
+        lastFailureReason: failureReason,
+        updatedAt: new Date().toISOString(),
```

Comment:
Would recommend using a single date so that these two times are logged identically.